pinot-common/src/test/java/org/apache/pinot/common/metrics/PinotPrometheusMetricsTest.java

```diff
@@ -0,0 +1,493 @@
+  protected PinotMetricsFactory _pinotMetricsFactory;
+
+  protected HttpClient _httpClient;
+
+  //config keys
+  private static final String CONFIG_KEY_JMX_EXPORTER_PARENT_DIR = "jmxExporterConfigsParentDir";
+  private static final String CONFIG_KEY_PINOT_METRICS_FACTORY = "pinotMetricsFactory";
+  private static final String CONFIG_KEY_CONTROLLER_CONFIG_FILE_NAME = "controllerConfigFileName";
+  private static final String CONFIG_KEY_SERVER_CONFIG_FILE_NAME = "serverConfigFileName";
+  private static final String CONFIG_KEY_BROKER_CONFIG_FILE_NAME = "brokerConfigFileName";
+  private static final String CONFIG_KEY_MINION_CONFIG_FILE_NAME = "minionConfigFileName";
+
+  //each meter defined in code is exported with these measurements
+  private static final List<String> METER_TYPES =
+      List.of("Count", "FiveMinuteRate", "MeanRate", "OneMinuteRate", "FifteenMinuteRate");
+
+  //each timer defined in code is exported with these measurements
+  private static final List<String> TIMER_TYPES =
+      List.of("Count", "FiveMinuteRate", "Max", "999thPercentile", "95thPercentile", "75thPercentile", "98thPercentile",
+          "OneMinuteRate", "50thPercentile", "99thPercentile", "FifteenMinuteRate", "Mean", "StdDev", "MeanRate",
+          "Min");
+
+  //each gauge defined in code is exported with these measurements
+  private static final List<String> GAUGE_TYPES = List.of("Value");
+
+  private String _exporterConfigsParentDir;
+
+  private final Map<PinotComponent, String> _pinotComponentToConfigFileMap = new HashMap<>();
+
+  @BeforeClass
+  public void setupTest()
+      throws Exception {
+    //read test configuration
+    JsonNode jsonNode = JsonUtils.DEFAULT_READER.readTree(loadResourceAsString("metrics/testConfig.json"));
+    _exporterConfigsParentDir = jsonNode.get(CONFIG_KEY_JMX_EXPORTER_PARENT_DIR).textValue();
+    _pinotComponentToConfigFileMap.put(PinotComponent.CONTROLLER,
+        jsonNode.get(CONFIG_KEY_CONTROLLER_CONFIG_FILE_NAME).textValue());
+    _pinotComponentToConfigFileMap.put(PinotComponent.SERVER,
+        jsonNode.get(CONFIG_KEY_SERVER_CONFIG_FILE_NAME).textValue());
+    _pinotComponentToConfigFileMap.put(PinotComponent.BROKER,
+        jsonNode.get(CONFIG_KEY_BROKER_CONFIG_FILE_NAME).textValue());
+    _pinotComponentToConfigFileMap.put(PinotComponent.MINION,
+        jsonNode.get(CONFIG_KEY_MINION_CONFIG_FILE_NAME).textValue());
+
+    String pinotMetricsFactory = jsonNode.get(CONFIG_KEY_PINOT_METRICS_FACTORY).textValue();
+    switch (pinotMetricsFactory) {
+      case "YammerMetricsFactory":
+        _pinotMetricsFactory = new YammerMetricsFactory();
+        break;
+      case "DropwizardMetricsFactory":
+        _pinotMetricsFactory = new DropwizardMetricsFactory();
+        break;
+      default:
+        throw new IllegalArgumentException("Unknow metrics factory specified in test config: " + pinotMetricsFactory
+            + ", supported ones are: YammerMetricsFactory and DropwizardMetricsFactory");
+    }
```

Comment:
I don't understand this code. Assuming in future we test both Yammer and Dropwizard... how do you expect we should write the tests? It looks it is reading some config from `testConfig.json` but it is not possible to read two different files so... why is this better than forcing to use Yammer in the code?

Reply:
Just to be clear. Due to the limitations of TestNG, what I would expect is to have 1 test class per component (broker, controller, etc) and per metric registry (Yammer, Dropwizard, etc).

In order to have so, I would expect PinotPrometheusMetricsTest to be abstract and have two abstract methods:
1. `PinotComponent getComponent()`, as explained in another comment.
2. `PinotMetricsFactory getPinotMetricsFactory()`, that returns the registry to be used.

Then for each component we may have another abstract class (like `BrokerPrometheusMetricsTest`) that inherits `PinotPrometheusMetricsTest` and implements `getComponents` and includes all test methods and finally two not-abstract classes `YammerBrokerPrometheusMetricsTest` and `DropwizardBrokerPrometheusMetricsTest` which extends `BrokerPrometheusMetricsTest` and only implements `getPinotMetricsFactory()`.

Something like:

```mermaid
classDiagram
  class PinotPrometheusMetricsTest {
      
  }
  class BrokerPrometheusMetricsTest {
      getComponent
  }
  class DropwizardBrokerPrometheusMetricsTest {
      getPinotMetricsFactory
  }
  class YammerBrokerPrometheusMetricsTest {
    getPinotMetricsFactory
  }
    class ServerPrometheusMetricsTest {
        getComponent
    }
    class DropwizardServerPrometheusMetricsTest {
        getPinotMetricsFactory
    }
    class YammerServerPrometheusMetricsTest {
        getPinotMetricsFactory
    }
  
  PinotPrometheusMetricsTest <|-- BrokerPrometheusMetricsTest
  BrokerPrometheusMetricsTest <|-- DropwizardBrokerPrometheusMetricsTest
  BrokerPrometheusMetricsTest <|-- YammerBrokerPrometheusMetricsTest

  PinotPrometheusMetricsTest <|-- ServerPrometheusMetricsTest
  ServerPrometheusMetricsTest <|-- DropwizardServerPrometheusMetricsTest
  ServerPrometheusMetricsTest <|-- YammerServerPrometheusMetricsTest
```

Reply:
Thanks for the detailed comment, have restructured the code.